Kube-proxy: ICMP reject via LBs when no endpoints

thockin opened the pull request:
/kind bug

**What this PR does / why we need it**:

ICMP reject services with no endpoints through LBs.  We already reject all other cases.

Fixes #48719

#72879 (by @marwinski) was almost identical, but I didn't see that until I had already written this.  Full credit to @marwinski .

**Does this PR introduce a user-facing change?**:

```release-note
Services of type=LoadBalancer which have no endpoints will now immediately ICMP reject connections, rather than time out.
```

thockin committed:
Kube-proxy: REJECT LB IPs with no endpoints

We REJECT every other case.  Close this FIXME.

To get this to work in all cases, we have to process service in
filter.INPUT, since LB IPS might be manged as local addresses.

danwinship commented on test/e2e/network/service.go:
```diff
@@ -791,11 +791,48 @@ var _ = SIGDescribe("Services", func() {
+		jig.Scale(ns1, 0)
+		jig.Scale(ns2, 0)
+
+		//FIXME: prove ICMP-reject
```
For the internal case (#72561) we ended up just assuming that we'd get "connection refused" rather than "timed out".

You could just check the counters on the iptables rules themselves; run `iptables-save -c` before and after, and verify that the packet count on the expected `REJECT` rule actually went up, and assume that if the right rule got hit, then the right end-user behavior happened too. (Though in fact, in this case the test would still pass even if the network plugin ate the ICMP reply...)

You could also check the global ICMP counters in `netstat -s`, which would let you confirm that the node had received an ICMP reject for *some* reason. That would have some unknowable number of false positives, but shouldn't have false negatives.

marwinski commented:
As this PR is virtually identical to #72879 I would appreciate if you name me as co-author and close the other one.

> #72879 is similar, but I didn't see it until I had already written this. The testing is the tricky part.

Testing is indeed tricky but here is how I did it manually (with some proposal to automate). Our problem was not the "connection refused" but the stale conntrack entries that were caused by not rejecting the packets. Downside is that it works only for providers that have load balancers that do not terminate the TCP connections. 

## Prepare service

```
> kubectl run nginx --image nginx --replicas=1
deployment.apps/nginx created
> kubectl expose deployment nginx --type=LoadBalancer --port=80
service/nginx exposed
> kubectl describe service  nginx
...
LoadBalancer Ingress:     104.40.197.150
> curl 104.40.197.150
<!DOCTYPE html>
<html>
...
```
### Log on to the host or a pod with host network (in this case calico-node)
 
Make sure there are not ```SYN_SENT``` entries:

```
> kubectl exec -it -n kube-system calico-node-bsn4f -c calico-node sh
/ # conntrack -L | grep SYN_SENT
conntrack v1.4.4 (conntrack-tools): 489 flow entries have been shown.
/ #
 ```
-> no stale SYN_SENT entries, all clean
 
## Start curl endless loop
 
```
> while true; do curl 104.40.197.150 ; done
<!DOCTYPE html>
<html>
...
```

### Scale replicas to 0
 
```
kubectl scale deployment nginx --replicas=0
deployment.extensions/nginx scaled
```
 
### Go to host and look for stale entries:

``` 
/ # conntrack -L | grep SYN_SENT
tcp      6 119 SYN_SENT src=109.192.86.132 dst=104.40.197.150 sport=60659 dport=80 [UNREPLIED] src=104.40.197.150 dst=109.192.86.132 sport=80 dport=60659 mark=0 secctx=system_u:object_r:unlabeled_t:s0 use=1
conntrack v1.4.4 (conntrack-tools): 434 flow entries have been shown.
/ #
```

With the fix we don't see any stale ```SYN_SENT``` entries. 

This can be simplified in the following way, however that works only if the load balancer is not a terminating one:

* create the service as shown above 
* Craft a connection request to the "external load balancer ip"  but with the mac address of a cluster node
* Without the rule you will see a hang when number of services is scaled to 0. You will also see stale SYN_SENT entries
* With the rule you should see a REJECT and no stale conntrack entries. 

This should be straightforward but will not work when NodePort is created for the LB (but is straightforward to automate).

nikopen commented:
/milestone v1.14

thockin commented:
Happy to name you as co-author (inasmuch as that's a thing).

We need to have automated testing for this.  I have been jumping between tasks, and have not had much time on this one.  I think I finally figured why the approach I had was not working.  If this run fails the way I expect, then we can proceed.

thockin commented:
/retest

nikopen commented:
/retest

thockin committed:
Retool HTTP and UDP e2e utils

This is a prefactoring for followup changes that need to use very
similar but subtly different test.  Now it is more generic, though it
pushes a little logic up the stack.  That makes sense to me.

thockin committed:
Fix small race in e2e

Occasionally we get spurious errors about "no route to host" when we
race with kube-proxy.  This should reduce that.  It's mostly just log
noise.

thockin commented:
OK, I think this will now pass tests.  I have enabled the slow service test here, just to prove it.  I will re-disable that for final commit.

PTAL @marwinski @danwinship

thockin commented:
/hold until I remove the FIXME (e2e test enabled)

thockin commented:
Tests passed with the service test enabled:

```
I0308 02:31:17.145] [1mSTEP[0m: Scaling the pods to 0
I0308 02:31:17.145] Mar  8 02:27:55.520: INFO: Waiting up to 2m0s for 0 pods to be created
I0308 02:31:17.145] Mar  8 02:27:55.573: INFO: Found all 0 pods
I0308 02:31:17.145] Mar  8 02:27:55.573: INFO: Waiting up to 2m0s for 0 pods to be running and ready: []
I0308 02:31:17.146] Mar  8 02:27:55.573: INFO: Wanted all 0 pods to be running and ready. Result: true. Pods: []
I0308 02:31:17.146] Mar  8 02:27:55.669: INFO: Waiting up to 2m0s for 0 pods to be created
I0308 02:31:17.146] Mar  8 02:27:55.732: INFO: Found all 0 pods
I0308 02:31:17.146] Mar  8 02:27:55.732: INFO: Waiting up to 2m0s for 0 pods to be running and ready: []
I0308 02:31:17.146] Mar  8 02:27:55.732: INFO: Wanted all 0 pods to be running and ready. Result: true. Pods: []
I0308 02:31:17.146] [1mSTEP[0m: looking for ICMP REJECT on the TCP service's NodePort
I0308 02:31:17.146] Mar  8 02:27:55.732: INFO: Poking "http://35.230.25.230:31513/"
I0308 02:31:17.146] Mar  8 02:28:25.732: INFO: Poke("http://35.230.25.230:31513/"): Get http://35.230.25.230:31513/: dial tcp 35.230.25.230:31513: i/o timeout
I0308 02:31:17.146] Mar  8 02:28:27.733: INFO: Poking "http://35.230.25.230:31513/"
I0308 02:31:17.147] Mar  8 02:28:27.772: INFO: Poke("http://35.230.25.230:31513/"): Get http://35.230.25.230:31513/: dial tcp 35.230.25.230:31513: connect: connection refused
I0308 02:31:17.147] [1mSTEP[0m: looking for ICMP REJECT on the UDP service's NodePort
I0308 02:31:17.147] Mar  8 02:28:27.772: INFO: Poking udp://35.230.25.230:30375
I0308 02:31:17.147] Mar  8 02:28:27.811: INFO: Poke("udp://35.230.25.230:30375"): read udp 10.61.15.82:46443->35.230.25.230:30375: read: connection refused
I0308 02:31:17.147] [1mSTEP[0m: looking for ICMP REJECT on the TCP service's LoadBalancer
I0308 02:31:17.147] Mar  8 02:28:27.811: INFO: Poking "http://35.230.125.171:81/"
I0308 02:31:17.147] Mar  8 02:28:27.851: INFO: Poke("http://35.230.125.171:81/"): Get http://35.230.125.171:81/: dial tcp 35.230.125.171:81: connect: connection refused
I0308 02:31:17.147] [1mSTEP[0m: looking for ICMP REJECT on the UDP service's LoadBalancer
I0308 02:31:17.148] Mar  8 02:28:27.851: INFO: Poking udp://35.233.206.44:81
I0308 02:31:17.148] Mar  8 02:28:27.891: INFO: Poke("udp://35.233.206.44:81"): read udp 10.61.15.82:33518->35.233.206.44:81: read: connection refused
I0308 02:31:17.148] [1mSTEP[0m: Scaling the pods to 1
```


I will now revert that part of the test.

thockin commented:
and it's pushed.

marwinski commented:
This is awesome, thanks

danwinship commented on test/e2e/framework/networking_utils.go:
```diff
@@ -911,13 +974,12 @@ func TestHitNodesFromOutsideWithCount(externalIP string, httpPort int32, timeout
-		if err != nil || !reached {
+		result := PokeHTTP(externalIP, int(httpPort), "/hostname", &HTTPPokeParams{Timeout: 1 * time.Second})
+		if result.Code != 0 {
 			return false, nil
```
This is wrong; it will return and retry on `result.Code == 200`, meaning there's no way to ever update the counts below and return `true`, meaning the function will always time out. (You probably meant `==` rather than `!=` here, although `if result.Status != HTTPOK` would be closer to the original code; you don't want to check the body if `Status == HTTPWrongCode`). 

It looks like this function only gets used from `It("[Slow] [Serial] should create valid firewall rules for LoadBalancer type service"` in `test/e2e/network/firewall.go`, which I guess isn't getting run by CI because it's Slow?

I guess you probably want to force a run with *all* Slow networking tests, to fully test the new code

thockin replied:
Nice catch.

danwinship commented on test/e2e/framework/service_util.go:
```diff
@@ -888,36 +898,60 @@ func (j *ServiceTestJig) TestReachableHTTPWithRetriableErrorCodes(host string, p
-		Failf("Could still reach UDP service through %v:%v after %v: %v", host, port, timeout, err)
+	pollfn := func() (bool, error) {
+		result := PokeUDP(host, port, "echo hello", &UDPPokeParams{Timeout: 3 * time.Second})
+		if result.Status != UDPOK && result.Status != UDPError {
```
This could potentially silently break if someone added a new `UDPPokeStatus` value in the future

thockin replied:
Yeah.  I though about characterizing failures but I don't know every possible network error that could be returned.  I figured since this is limited scope, it might be OK.  What if I just add a comment at the definition of these codes saying any additions here need to audit callers of this.  Good enough?

danwinship replied:
sure

danwinship commented on test/e2e/framework/service_util.go:
```diff
@@ -888,36 +898,60 @@ func (j *ServiceTestJig) TestReachableHTTPWithRetriableErrorCodes(host string, p
-			return false, nil
+		result := PokeHTTP(host, port, url, nil)
+		if result.Code != 0 {
+			body.Write(result.Body)
```
This will return the body even if it gets a 404 or 500 response, which is a change from the old code. It should be `if result.Status == HTTPOK`

thockin replied:
Agree

danwinship commented on test/e2e/network/service.go:
```diff
@@ -2063,14 +2063,17 @@ var _ = SIGDescribe("ESIPP [Slow] [DisabledForLargeClusters]", func() {
+				if result.Code == 0 {
+					return true, nil
+				}
+				body.Write(result.Body)
```
If the poll fails more than once, this will end up concatenating multiple response bodies together, as opposed to the old code which always only returned the last HTTP response

thockin replied:
nice catch

danwinship commented on test/e2e/framework/networking_utils.go:
```diff
@@ -708,197 +707,261 @@ func CheckReachabilityFromPod(expectToBeReachable bool, timeout time.Duration, n
+	ret.Response = buf
 
-	return false, nil
+	if params.ResponseContains != "" && !strings.Contains(string(buf), params.ResponseContains) {
```
It only read `len(params.ResponseContains)` bytes into `buf`, so this isn't really `ResponseContains`, it's `ResponsePrefix`.

But maybe (matching the assumptions on the writing side) you should rename it to just `Response`, and then pass a buffer of length `len(params.Response)+2` to `Read()`, and check that the returned length is `len(params.Response)+1` and that it matches `params.Response+"\n"`. (ie, the server sent back exactly `Response` plus a newline, and nothing more.)

thockin replied:
Why bother with newline, just do exact comparison?  Will fix

danwinship replied:
Oh, oops... the code appends a newline to the request string, and the original `TestReachableUDP` read back one byte more than the length of the response string but then didn't look at the last byte, so I was assuming that there was a newline in the response too. But I guess there isn't. (And, in fact, the server doesn't care if there's a newline in the request either; it just ignores any trailing whitespace.)

thockin commented:
I will process the feedback and re-run tests.  Great points.  Thanks.

thockin commented on test/e2e/framework/networking_utils.go:
```diff
@@ -708,197 +707,261 @@ func CheckReachabilityFromPod(expectToBeReachable bool, timeout time.Duration, n
-
-func TestReachableHTTPWithContentTimeoutWithRetriableErrorCodes(ip string, port int, request string, expect string, content *bytes.Buffer, retriableErrCodes []int, timeout time.Duration) (bool, error) {
+const (
+	HTTPOK          HTTPPokeStatus = "OK"
```
OK

j/k - changed to "success" in next push

thockin reviewed (commented):
Excellent code review!  Thank you.

I will be pushing a new tree momentarily, with a bunch of Slow test enabled.

thockin commented:
/hold until slow tests are reverted

thockin commented:
Tests pass.  Will revert the test-enable changes

thockin commented:
I have re-pushed.  PTAL.

danwinship commented:
/lgtm

thockin commented:
/retest

thockin commented:
/retest

thockin commented:
@m1093782566 Can you take a look at IPVS and see if it suffers similarly (https://github.com/kubernetes/kubernetes/issues/48719 is the symptom)

thockin commented:
@m1093782566 please?

m1093782566 commented:
@thockin 

I don't think IPVS has the same issue.
